# Lucid Software Analytics Case Study

In [510]:
import numpy as np
import pandas as pd
paywall = pd.read_csv('/Users/LSP/Downloads/lucidanalyticsexercise/paywall_data.csv')
payment = pd.read_csv('/Users/LSP/Downloads/lucidanalyticsexercise/paywall_payment_data.csv')
result = pd.merge(paywall,payment, on='id', how='left')

In [511]:
for i in result:
    if i not in ['id', 'Paid?', 'First Payment Value']:
        result[i] = pd.to_datetime(result[i])

In [670]:
no_hit_paywall =pd.Series(result.count()[3:-2])
no_hit_paywall

(Pw) Analytics                    270
(Pw) Backup-restore                64
(Pw) Color-Management            2838
(Pw) Custom-Fonts                2025
(Pw) Downloadable-Publication    1501
(Pw) Email-Thumb                 1260
(Pw) Embedding                    195
(Pw) Page-Limit                  3519
(Pw) Pdf-dpi                     4268
(Pw) Premium Templates           3400
(Pw) Pub-Password                 297
(Pw) Publish-banner               725
(Pw) Revision-history             141
(Pw) Storage                     2786
(Pw) Unlicensed/View Only         132
dtype: int64

In [589]:
result1 = result[result['Paid?'] ==1]

In [590]:
hhh=[]
ddd=[]
for i in range(0,2035):
    a= result1.iloc[i,3:-1].sort_values(axis=0,ascending =False)
    if a[0] == a['First Payment Date']:
        hhh.append(a.index[1])
        ddd.append(a[1])
    elif a[1] == a['First Payment Date']:
        hhh.append(a.index[2])
        ddd.append(a[2])
    elif a[2] == a['First Payment Date']:
        hhh.append(a.index[3])
        ddd.append(a[3])
    elif a[3] == a['First Payment Date']:
        hhh.append(a.index[4])
        ddd.append(a[4])
    elif a[4] == a['First Payment Date']:
        hhh.append(a.index[5])
        ddd.append(a[5])
    elif a[5] == a['First Payment Date']:
        hhh.append(a.index[6])
        ddd.append(a[6])
    elif a[6] == a['First Payment Date']:
        hhh.append(a.index[7])
        ddd.append(a[7])
    elif a[7] == a['First Payment Date']:
        hhh.append(a.index[8])
        ddd.append(a[8])
    elif a[8] == a['First Payment Date']:
        hhh.append(a.index[9])
        ddd.append(a[9])
    elif a[9] == a['First Payment Date']:
        hhh.append(a.index[10])
        ddd.append(a[10])
    else:
        hhh.append('ERROR')
        ddd.append('ERROR')




In [591]:
result1['paywall_res']= hhh
result1['paywall_date']= ddd

/Users/LSP/opt/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  """Entry point for launching an IPython kernel.
/Users/LSP/opt/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  


In [721]:
result1.iloc[0,3:-3].sort_values(axis=0,ascending =False)

(Pw) Downloadable-Publication   2016-02-20 10:04:00
First Payment Date              2016-02-20 09:57:00
(Pw) Page-Limit                 2016-02-20 09:37:00
(Pw) Storage                    2016-02-13 08:54:00
(Pw) Analytics                                  NaT
(Pw) Backup-restore                             NaT
(Pw) Color-Management                           NaT
(Pw) Custom-Fonts                               NaT
(Pw) Email-Thumb                                NaT
(Pw) Embedding                                  NaT
(Pw) Pdf-dpi                                    NaT
(Pw) Premium Templates                          NaT
(Pw) Pub-Password                               NaT
(Pw) Publish-banner                             NaT
(Pw) Revision-history                           NaT
(Pw) Unlicensed/View Only                       NaT
Name: 53228, dtype: datetime64[ns]

In [582]:
result1.iloc[6,3:-2].sort_values(axis=0,ascending =False)

(Pw) Color-Management           2016-05-12 00:54:00
(Pw) Downloadable-Publication   2016-04-25 13:22:00
(Pw) Storage                    2016-04-21 14:10:00
(Pw) Page-Limit                 2016-04-19 01:42:00
(Pw) Custom-Fonts               2016-03-08 22:38:00
First Payment Date              2016-02-26 23:16:00
(Pw) Analytics                                  NaT
(Pw) Backup-restore                             NaT
(Pw) Email-Thumb                                NaT
(Pw) Embedding                                  NaT
(Pw) Pdf-dpi                                    NaT
(Pw) Premium Templates                          NaT
(Pw) Pub-Password                               NaT
(Pw) Publish-banner                             NaT
(Pw) Revision-history                           NaT
(Pw) Unlicensed/View Only                       NaT
Name: 53234, dtype: datetime64[ns]

In [718]:
result1.iloc[:10,3:]

(Pw) Analytics (Pw) Backup-restore (Pw) Color-Management  \
53228            NaT                 NaT                   NaT   
53229            NaT                 NaT                   NaT   
53230            NaT                 NaT                   NaT   
53231            NaT                 NaT                   NaT   
53232            NaT                 NaT                   NaT   
53233            NaT                 NaT                   NaT   
53234            NaT                 NaT   2016-05-12 00:54:00   
53235            NaT                 NaT                   NaT   
53236            NaT                 NaT                   NaT   
53237            NaT                 NaT                   NaT   

        (Pw) Custom-Fonts (Pw) Downloadable-Publication (Pw) Email-Thumb  \
53228                 NaT           2016-02-20 10:04:00              NaT   
53229                 NaT                           NaT              NaT   
53230 2016-05-01 08:37:00           2016-05-05 02:04:00              NaT   
53231                 NaT                           NaT              NaT   
53232 2016-01-07 15:52:00                           NaT              NaT   
53233                 NaT                           NaT              NaT   
53234 2016-03-08 22:38:00           2016-04-25 13:22:00              NaT   
53235                 NaT                           NaT              NaT   
53236                 NaT                           NaT              NaT   
53237                 NaT                           NaT              NaT   

      (Pw) Embedding     (Pw) Page-Limit        (Pw) Pdf-dpi  \
53228            NaT 2016-02-20 09:37:00                 NaT   
53229            NaT 2016-04-16 02:50:00                 NaT   
53230            NaT 2016-04-17 08:41:00                 NaT   
53231            NaT 2016-02-11 14:46:00                 NaT   
53232            NaT 2016-01-14 16:26:00                 NaT   
53233            NaT                 NaT 2016-03-08 09:20:00   
53234            NaT 2016-04-19 01:42:00                 NaT   
53235            NaT 2016-03-31 00:59:00                 NaT   
53236            NaT 2016-03-07 14:05:00 2016-03-09 17:34:00   
53237            NaT 2016-04-10 22:15:00 2016-05-22 22:54:00   

      (Pw) Premium Templates (Pw) Pub-Password (Pw) Publish-banner  \
53228                    NaT               NaT                 NaT   
53229    2016-04-15 21:59:00               NaT                 NaT   
53230                    NaT               NaT                 NaT   
53231                    NaT               NaT                 NaT   
53232                    NaT               NaT 2016-01-14 16:14:00   
53233                    NaT               NaT                 NaT   
53234                    NaT               NaT                 NaT   
53235                    NaT               NaT                 NaT   
53236    2016-03-06 08:13:00               NaT                 NaT   
53237                    NaT               NaT                 NaT   

      (Pw) Revision-history        (Pw) Storage (Pw) Unlicensed/View Only  \
53228                   NaT 2016-02-13 08:54:00                       NaT   
53229                   NaT                 NaT                       NaT   
53230                   NaT 2016-04-28 15:31:00                       NaT   
53231                   NaT                 NaT                       NaT   
53232                   NaT                 NaT                       NaT   
53233                   NaT                 NaT                       NaT   
53234                   NaT 2016-04-21 14:10:00                       NaT   
53235                   NaT                 NaT                       NaT   
53236                   NaT                 NaT                       NaT   
53237                   NaT 2016-05-17 23:19:00                       NaT   

       First Payment Date  First Payment Value      paywall_res  \
53228 2016-02-20 09:57:00                 9.95  (Pw

In [723]:
result1[result1['paywall_date'].isna()].paywall_res.value_counts()

(Pw) Analytics         840
(Pw) Backup-restore     36
Name: paywall_res, dtype: int64

In [629]:
result1[result1.paywall_res == '(Pw) Analytics'].shape

(858, 22)

In [630]:
result1[result1.paywall_res == '(Pw) Backup-restore'].shape

(43, 22)

In [632]:
result2 = result1[result1['paywall_date'].notna()].iloc[:,2:]

In [672]:
rev= pd.Series(result2.groupby('paywall_res').sum()['First Payment Value'],index=None)

paywall_res
(Pw) Analytics                    1143.01
(Pw) Backup-restore                294.00
(Pw) Color-Management             4450.42
(Pw) Custom-Fonts                 4406.14
(Pw) Downloadable-Publication     2806.78
(Pw) Email-Thumb                  2133.25
(Pw) Embedding                    1030.25
(Pw) Page-Limit                  10748.02
(Pw) Pdf-dpi                     10677.58
(Pw) Premium Templates            2323.54
(Pw) Pub-Password                  306.65
(Pw) Publish-banner                435.18
(Pw) Revision-history               39.85
(Pw) Storage                      8143.80
(Pw) Unlicensed/View Only          111.30
Name: First Payment Value, dtype: float64

In [673]:
no_upgrade = pd.Series(result2.groupby('paywall_res').count()['paywall_date'],index=None)

In [700]:
result_output =pd.DataFrame({'Total no. of hit': no_hit_paywall, 'No. of upgrade':no_upgrade, \
                             'Conversion rate': no_upgrade/no_hit_paywall, 'Revenue': rev},index=None)

# 1. Which paywalls have the best conversion rate?
Note: Conversion rate is defined as the number of people who pay because of a given paywall
divided by the number of people who hit the paywall. An important part of this task is deciding
how you will determine whether a paywall caused a user to pay or not. Also note that a user can
continue to hit paywalls after their first payment for features that are not part of their subscription
plan (Lucidpress has multiple subscription plans as shown here ).

In [702]:
result_output.sort_values('Conversion rate',ascending = False)

Total no. of hit  No. of upgrade  \
paywall_res                                                       
(Pw) Backup-restore                          64               7   
(Pw) Page-Limit                            3519             331   
(Pw) Pdf-dpi                               4268             292   
(Pw) Storage                               2786             189   
(Pw) Analytics                              270              18   
(Pw) Embedding                              195              10   
(Pw) Custom-Fonts                          2025              74   
(Pw) Email-Thumb                           1260              39   
(Pw) Downloadable-Publication              1501              46   
(Pw) Color-Management                      2838              84   
(Pw) Unlicensed/View Only                   132               3   
(Pw) Revision-history                       141               3   
(Pw) Pub-Password                           297               6   
(Pw) Publish-banner                         725              13   
(Pw) Premium Templates                     3400              44   

                               Conversion rate   Revenue  
paywall_res                                               
(Pw) Backup-restore                   0.109375    294.00  
(Pw) Page-Limit                       0.094061  10748.02  
(Pw) Pdf-dpi                          0.068416  10677.58  
(Pw) Storage                          0.067839   8143.80  
(Pw) Analytics                        0.066667   1143.01  
(Pw) Embedding                        0.051282   1030.25  
(Pw) Custom-Fonts                     0.036543   4406.14  
(Pw) Email-Thumb                      0.030952   2133.25  
(Pw) Downloadable-Publication         0.030646   2806.78  
(Pw) Color-Management                 0.029598   4450.42  
(Pw) Unlicensed/View Only             0.022727    111.30  
(Pw) Revision-history                 0.021277     39.85  
(Pw) Pub-Password                     0.020202    306.65  
(Pw) Publish-banner                   0.017931    435.18  
(Pw) Premium Templates                0.012941   2323.54

Backup-restore, Page-Limit,  Pdf-dpi, Storage, and Analytics are the top 5 features with the highest conversion rates.


**(Note: 
The dataset only gives first payment date. There could be cases that a user chooses to upgrade from "Pro" to "Team" on a second payment because they hit a new paywall. Therefore, the conversion rate for those extra features in the "Team" package may not be accurate.
In addition, information about subscription period is missing. There seem cases that a user has subscribed to a paid plan but encountered paywalls because by the time they used the features their suscription had expired. This may further jeopardize the validity of the claim.)**

# 2. Which paywalls contribute most to revenue?
Note: Be aware that revenue can be different for each customer because Lucidpress has
different subscription plans, as noted above.

In [703]:
result_output.sort_values('Revenue',ascending = False)

Total no. of hit  No. of upgrade  \
paywall_res                                                       
(Pw) Page-Limit                            3519             331   
(Pw) Pdf-dpi                               4268             292   
(Pw) Storage                               2786             189   
(Pw) Color-Management                      2838              84   
(Pw) Custom-Fonts                          2025              74   
(Pw) Downloadable-Publication              1501              46   
(Pw) Premium Templates                     3400              44   
(Pw) Email-Thumb                           1260              39   
(Pw) Analytics                              270              18   
(Pw) Embedding                              195              10   
(Pw) Publish-banner                         725              13   
(Pw) Pub-Password                           297               6   
(Pw) Backup-restore                          64               7   
(Pw) Unlicensed/View Only                   132               3   
(Pw) Revision-history                       141               3   

                               Conversion rate   Revenue  
paywall_res                                               
(Pw) Page-Limit                       0.094061  10748.02  
(Pw) Pdf-dpi                          0.068416  10677.58  
(Pw) Storage                          0.067839   8143.80  
(Pw) Color-Management                 0.029598   4450.42  
(Pw) Custom-Fonts                     0.036543   4406.14  
(Pw) Downloadable-Publication         0.030646   2806.78  
(Pw) Premium Templates                0.012941   2323.54  
(Pw) Email-Thumb                      0.030952   2133.25  
(Pw) Analytics                        0.066667   1143.01  
(Pw) Embedding                        0.051282   1030.25  
(Pw) Publish-banner                   0.017931    435.18  
(Pw) Pub-Password                     0.020202    306.65  
(Pw) Backup-restore                   0.109375    294.00  
(Pw) Unlicensed/View Only             0.022727    111.30  
(Pw) Revision-history                 0.021277     39.85

Page-Limit,  Pdf-dpi, Storage, Color-Management, and Custom-Fonts are the top 5 features with the highest revenues.

**(Note: there are a total of 2035 cases of payment. 1159 of them are due to hitting a paywall, while 876 of them are upgrades before hitting a paywall. These 876 are taken out when summing the total revenue by each paywall.)**

In [726]:
result1[result1['paywall_date'].isna()][result1['paywall_res']=='(Pw) Backup-restore']

/Users/LSP/opt/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:1: UserWarning: Boolean Series key will be reindexed to match DataFrame index.
  """Entry point for launching an IPython kernel.


id  Paid?   Registration Date      (Pw) Analytics  \
53278   9428212      1 2016-01-03 13:23:00 2016-01-18 11:31:00   
53285   8772107      1 2016-01-06 18:07:00 2016-06-06 11:01:00   
53329  10804625      1 2016-01-26 23:50:00 2016-05-26 23:19:00   
53344  11020447      1 2016-02-01 18:47:00 2016-04-26 19:19:00   
53361  11231533      1 2016-02-08 09:58:00 2016-05-31 14:06:00   
53409  11891898      1 2016-02-28 01:25:00 2016-03-30 02:41:00   
53418  12081300      1 2016-03-03 13:22:00 2016-05-24 17:28:00   
53443  12400805      1 2016-03-13 17:17:00 2016-04-26 08:17:00   
53487  12919185      1 2016-03-30 11:00:00 2016-04-14 10:09:00   
53603  10924675      1 2016-01-29 00:47:00 2016-04-10 14:43:00   
53668  10579439      1 2016-01-18 01:46:00 2016-01-26 03:43:00   
53778  10976860      1 2016-01-31 14:32:00 2016-04-08 15:24:00   
53779   3736796      1 2016-01-27 10:02:00 2016-03-10 10:36:00   
53841  13012397      1 2016-03-31 16:20:00 2016-05-08 14:41:00   
53874  11063745      1 2016-02-03 07:02:00 2016-02-08 08:27:00   
53888  11240875      1 2016-02-08 13:14:00 2016-04-12 11:10:00   
53916  10320565      1 2016-01-07 12:26:00 2016-01-13 12:24:00   
53983   4439781      1 2015-02-03 01:44:00 2015-03-26 19:57:00   
54133  11844421      1 2016-02-26 01:20:00 2016-03-13 22:23:00   
54164  12308979      1 2016-03-10 20:28:00 2016-03-11 17:04:00   
54272  11567028      1 2016-02-18 07:04:00 2016-02-25 09:27:00   
54308  12218422      1 2016-03-08 07:38:00 2016-04-06 09:49:00   
54359  10675156      1 2016-01-21 00:33:00 2016-01-27 10:10:00   
54478  11334930      1 2016-02-10 15:13:00 2016-02-17 18:48:00   
54489   2842513      1 2014-10-02 03:47:00 2015-09-28 00:54:00   
54602  10185469      1 2016-01-01 10:41:00 2016-01-19 17:19:00   
54629  11364964      1 2016-02-11 10:01:00 2016-02-18 06:59:00   
54691   4504633      1 2015-02-05 21:41:00 2015-07-14 00:10:00   
54709  12652673      1 2016-03-21 11:51:00 2016-03-29 15:32:00   
54844  12197972      1 2016-03-07 16:38:00 2016-03-11 13:58:00   
54865  11132366      1 2016-02-04 10:10:00 2016-02-25 12:35:00   
54877   6011035      1 2015-05-05 21:49:00 2016-03-10 08:32:00   
54938  11300216      1 2016-03-28 13:25:00 2016-04-27 23:01:00   
54946  11945011      1 2016-02-29 16:20:00 2016-03-03 09:42:00   
55225   4824789      1 2015-07-08 19:45:00 2015-08-04 16:44:00   
55244   3695938      1 2016-02-23 20:26:00 2015-06-09 22:46:00   

      (Pw) Backup-restore (Pw) Color-Management   (Pw) Custom-Fonts  \
53278                 NaT   2016-03-08 16:06:00 2016-02-04 15:28:00   
53285                 NaT   2016-06-13 12:37:00 2016-06-04 15:41:00   
53329                 NaT   2016-05-26 13:14:00 2016-01-28 00:32:00   
53344                 NaT   2016-04-25 18:17:00 2016-02-25 22:57:00   
53361                 NaT   2016-05-31 14:09:00                 NaT   
53409                 NaT   2016-05-30 16:54:00 2016-05-30 10:24:00   
53418                 NaT   2016-03-31 10:50:00                 NaT   
53443                 NaT   2016-05-05 03:39:00 2016-03-29 05:26:00   
53487                 NaT                   NaT                 NaT   
53603                 NaT   2016-04-22 00:27:00                 NaT   
53668                 NaT   2016-01-26 03:39:00                 NaT   
53778                 NaT                   NaT                 NaT   
53779                 NaT                   NaT                 NaT   
53841                 NaT   2016-05-08 18:26:00                 NaT   
53874                 NaT   2016-05-11 12:56:00                 NaT   
53888                 NaT                   NaT                 NaT   
53916                 NaT   2016-02-02 11:47:00                 NaT   
53983                 NaT   2016-03-22 16:01:00 2016-03-30 11:31:00   
54133                 NaT                   NaT                 NaT   
54164                 NaT   2016-03-11 21:55:00                 NaT   
54272                 NaT   2016-03-01 09:14:00                 NaT   
54308   

# 3. Methodology


To answer question #1, 
1. I first inner-joined the payment dataset with the paywall dataset to get the first payment date;
2. I converted datatypes from strings to datetime in relevant columns to compare the datetimes of hitting a wall with first payment date;
3. Assuming that a user chooses to upgrade because of the last paywall they hit (while an alternative could be the user considers previous paywalls and the last paywall is just the last straw), I attributed that payment to the last paywall hit;
4. While experimenting with the dataset, I discovered that some users upgraded without hitting a paywall. As a result, there are 840 cases and 36 cases wrongly attributed to Analytics and Backup-restore. These cases were excluded from the analysis afterward;
5. I grouped the dataset by paywall to get the no. of upgrade due to each paywall;
6. I found the total no. of hitting a paywall by counting the no. of timestamp under that paywall;
7. I calculated the conversion rate by dividing no. of upgrade by total no. of hit.


To answer question #2,
1. I excluded "auto-upgrade" cases when calculating revenue;
2. I grouped the dataset by paywall and summed up the First Payment Value in each category.


# 4. Recommendations for product changes


1. while Color-Management, Custom-Fonts, Downloadable-Publication, Premium Templates, and Email-Thumb are popular paywalls (with over 1,000 hits), they have relatively low conversion rates (below 0.04). These features could be potential revenue generators. I suggest marketing these features more aggressively to increase the necessity of having these features.

2. Page-Limit and Pdf-dpi are top 2 most frequently hit paywalls. A new one-time payment (e.g. 99 cents for a one-time unlimited page use) could be introduced to entice users to try these features without paying for the full amount of a package deal.

# 5. Recommendations for further analysis

1. We could track whether a subscriber chooses to further upgrade their plan and the duration for their subscription to understand the marketability of the premium features.

2. For popular features like Page-Limit, Pdf-dpi, Storage, Color-Management, Custom-Fonts, Downloadable-Publication, Premium Templates, and Email-Thumb, we could collect more data on the user profiles to better identify user personas for targeted marketing.

3. We could also analyze the time gaps between hitting a paywall and making a payment and introduce a notification feature or promotion to entice actions.